### DISCUSSION TITLE
Building Strimzi from source

### DISCUSSION THREAD

**Discussion by john-mcpeek** [strimzi/strimzi-kafka-operator#9532] [General]
I want to add a `custom` `sasl.mechanism` to the topic operator, but I haven't been able to get the code to build reliably. I followed the steps in the [Dev Guide](https://github.com/strimzi/strimzi-kafka-operator/blob/main/development-docs/DEV_GUIDE.md).  The change to the Topic Operator config seems simple, but if I can't get the tests to run reliably, I can't know if it works.

I run the `make all` in a Ubuntu terminal with Kubernetes running in Docker Desktop. The errors I get are timeouts; outputs below. Any thoughts on what I can look for?

Thanks.

This is my system:
Windows 11
#### Ubuntu 22.04.3 LTS (WSL2)
ZSH (also tried bash)
Java version: 17.0.9
make - GNU Make 4.3
Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)
Asciidoctor 2.0.16
Asciidoctor PDF 2.3.10 using Asciidoctor 2.0.16
yq (https://github.com/mikefarah/yq/) version v4.40.5
Docker - 24.0.6
ShellCheck - version: 0.8.0
Kubernetes - v1.28.2 (Docker desktop)

```
[ERROR] Failures:
[ERROR]   KafkaBridgeCrdOperatorIT.testUpdateStatusAfterResourceDeletedThrowsKubernetesClientException(VertxTestContext) java.lang.AssertionError:
Expected: an instance of io.fabric8.kubernetes.client.KubernetesClientException
     but: <io.strimzi.operator.common.operator.resource.TimeoutException: Exceeded timeout of 10000ms while waiting for KafkaBridge resource my-test-resource-1953004668 in namespace bridge-crd-it-namespace to be ready> is a io.strimzi.operator.common.operator.resource.TimeoutException
[ERROR]   KafkaConnectCrdOperatorIT.testUpdateStatusAfterResourceDeletedThrowsKubernetesClientException(VertxTestContext) java.lang.AssertionError:
Expected: an instance of io.fabric8.kubernetes.client.KubernetesClientException
     but: <io.strimzi.operator.common.operator.resource.TimeoutException: Exceeded timeout of 10000ms while waiting for KafkaConnect resource my-test-resource-1615204214 in namespace kafka-connect-crd-it-namespace to be ready> is a io.strimzi.operator.common.operator.resource.TimeoutException
[ERROR]   KafkaCrdOperatorIT.testUpdateStatusAfterResourceDeletedThrowsKubernetesClientException(VertxTestContext) java.lang.AssertionError:
Expected: an instance of io.fabric8.kubernetes.client.KubernetesClientException
     but: <io.strimzi.operator.common.operator.resource.TimeoutException: Exceeded timeout of 10000ms while waiting for Kafka resource my-test-resource-166642516 in namespace kafka-crd-it-namespace to be ready> is a io.strimzi.operator.common.operator.resource.TimeoutException
[ERROR]   KafkaMirrorMaker2CrdOperatorIT.testUpdateStatusAfterResourceDeletedThrowsKubernetesClientException(VertxTestContext) java.lang.AssertionError:
Expected: an instance of io.fabric8.kubernetes.client.KubernetesClientException
     but: <io.strimzi.operator.common.operator.resource.TimeoutException: Exceeded timeout of 10000ms while waiting for KafkaMirrorMaker2 resource my-test-resource-480202737 in namespace kafka-mirror-maker-2-crd-it-namespace to be ready> is a io.strimzi.operator.common.operator.resource.TimeoutException
[ERROR]   KafkaMirrorMakerCrdOperatorIT.testUpdateStatusAfterResourceDeletedThrowsKubernetesClientException(VertxTestContext) java.lang.AssertionError:
Expected: an instance of io.fabric8.kubernetes.client.KubernetesClientException
     but: <io.strimzi.operator.common.operator.resource.TimeoutException: Exceeded timeout of 10000ms while waiting for KafkaMirrorMaker resource my-test-resource-549826474 in namespace kafka-mirror-maker-2-crd-it-namespace to be ready> is a io.strimzi.operator.common.operator.resource.TimeoutException
[ERROR]   StrimziPodSetCrdOperatorIT.testUpdateStatusAfterResourceDeletedThrowsKubernetesClientException(VertxTestContext) java.lang.AssertionError:
Expected: an instance of io.fabric8.kubernetes.client.KubernetesClientException
     but: <io.strimzi.operator.common.operator.resource.TimeoutException: Exceeded timeout of 10000ms while waiting for StrimziPodSet resource my-test-resource-1006495911 in namespace strimzipodset-crd-it-namespace to be ready> is a io.strimzi.operator.common.operator.resource.TimeoutException
[INFO]
[ERROR] Tests run: 34, Failures: 6, Errors: 0, Skipped: 0
[INFO]
[INFO]
[INFO] --- maven-failsafe-plugin:3.1.2:verify (default) @ operator-common ---
```

![image](https://github.com/strimzi/strimzi-kafka-operator/assets/6014230/e5e02b6c-6274-4d75-a18d-74c575bbfcdc)

**Comment by scholzj** [comment]
Unfortunately, these tests are sometimes flaky depending on how quickly your Kubernetes cluster does things. We just merged a PR (https://github.com/strimzi/strimzi-kafka-operator/pull/8715) that tries to make them more stable today morning. I think it among other things increases some of the timeouts. So maybe you can try to pull that and see if it helps. You can also skip the tests => `MVN_ARGS=-DskipITs make all` for integration tests or `MVN_ARGS=-DskipTests make all` for all tests. If you change only the Topic Operator, you can then run the tests only for it separately.

**Reply by john-mcpeek** [reply]
I am working on the change I was talking about. I ran into an issue that looks like a missing test to me. Is this this place to talk or is there a better slack channel?

**Reply by scholzj** [reply]
I think this place is as good as Slack is.

**Reply by john-mcpeek** [reply]
Ok,

ConfigTest#testInvalidSaslConfig fails after my change because I am using the SASL_MECHANISM to pick normal or custom sasl config. I use a switch with a default of `throw new IllegalArgumentException("Invalid SASL_MECHANISM type: " + configSaslMechanism);`. However, that test doesn't set a `SASL_MECHANISM` and I think we should test for that. I want to change the test to also test for the `SASL_MECHANISM` because if SASL is enabled, the mechanism should also be require. What do you think?

**Reply by scholzj** [reply]
@fvaleri @tombentley Any ideas? ^^^

**Reply by fvaleri** [reply]
Hi @john-mcpeek. The Strimzi doc says that SASL_ENABLED is optional and false by default. When it is set to true (enabled), all other STRIMZI_SASL_* configs are mandatory.

ConfigTest#testInvalidSaslConfig looks correct to me. It expects a failure if you enable the SASL authentication without any additional configuration, then it expects a failure if you add username and password, but not the mechanism.

**Reply by john-mcpeek** [reply]
BTW, thank for the feedback.

**Comment by john-mcpeek** [comment]
Ok, I need some help. I'm trying to push a branch and I'm getting access denied. I have pat for my repos, but I'm guessing I need access from you guys to even push the branch for a PR.? Or am I just missing something else?

**Reply by scholzj** [reply]
If you want to open a PR, you have to fork the Strimzi repository, push the branch into your own fork, and then open the PR.
